frontend/src/components/special/CopyIndicator.tsx

```diff
@@ -28,31 +31,92 @@ export const BottomCopyIndicatorContainer = styled.div.attrs((props: CopyIndicat
   bottom: 20px;
   left: 50%;
   transition: transform 0.25s;
+  line-height: 0;
+  z-index: 5;
 `;
 
-export const CopyIndicator = styled(DefaultButton)`
+const CopyIndicator = styled(DefaultButton)`
   position: relative;
   left: -50%;
   margin: 0 auto;
   padding: 0.25rem 1rem;
+  line-height: normal;
   color: ${({ theme }) => theme.colors.white};
   background: ${({ theme }) => theme.colors.gradients.green};
 `;
 
-export const InlineBackgroundCopyText = styled(ContactHeaderText)`
-  display: inline-block;
-  margin: 0;
-  padding: 0.25rem 0.5rem;
-  font-size: ${({ theme }) => theme.fontSize.mediumLarge};
-  background: ${({ theme }) => theme.colors.text};
-  color: ${({ theme }) => theme.colors.white};
-  border-radius: 0.5rem;
-  cursor: pointer;
-`;
-
-export const InlineCopyIcon = styled.i.attrs(() => ({
-  className: 'material-icons',
-}))`
+const InlineCopyIconWrapper = styled.i`
   margin-left: 5px;
   font-size: inherit;
 `;
+
+export const InlineCopyIcon = () => (
+  <InlineCopyIconWrapper className="material-icons">
+    content_copy
+  </InlineCopyIconWrapper>
+);
+
+type CopyableContentProps = {
+  children: React.ReactNode,
+  text: string,
+  top: boolean,
+};
+
+type CopyableProps = {
+  text: string,
+  top: boolean,
+};
+
+export function CopyableContent(props: CopyableContentProps) {
+  const { children, text, top } = props;
+
+  const [copied, setCopied] = useState(false);
+  const timeoutRef = useRef<ReturnType<typeof setTimeout> | null>(null);
+  const Container = top ? CopyIndicatorContainer : BottomCopyIndicatorContainer;
+
+  const onCopy = () => {
+    copy(text);
+    setCopied(true);
+
+    if (timeoutRef.current) {
+      clearTimeout(timeoutRef.current);
+    }
+    timeoutRef.current = setTimeout(() => setCopied(false), 2000);
+  };
+
+  const closeIndicator = () => {
+    setCopied(false);
+
+    if (timeoutRef.current) {
+      clearTimeout(timeoutRef.current);
+      timeoutRef.current = null;
+    }
+  };
+
+  return (
+    <>
+      <Container copied={copied}>
+        <CopyIndicator onClick={closeIndicator}>
+          Copied!&nbsp;&nbsp;✕
+        </CopyIndicator>
+      </Container>
+
+      <InlineContainer onClick={onCopy}>
+        {children}
+      </InlineContainer>
+    </>
+  );
+}
+
+export function Copyable(props: CopyableProps) {
```

Comment:
Actually out of curiosity, can you tell me how you implemented the two-second feature? I saw the code with the `useRef` but I'm not very familiar with that hook.

Reply:
Basically setTimeout returns some kind of identifier/value that you can use to reference the timeout - using this, anytime the button is clicked, it will first check whether a previous timeout is already scheduled (with useRef) and cancel it if so before scheduling a new timeout. 

The reason I used useRef to store the returned identifiers was actually cause of something you mentioned haha - from what I understood in that convo it's similar to useState in terms of storing values but doesn't trigger a re-render when the value changes